Refactor codec specs into a single doc (#102)
zarr-developers/zarr-specs: docs/codecs.rst

descriptive assertions and [RFC2119]_ terminology. The key words
"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in the normative
parts of this document are to be interpreted as described in
[RFC2119]_. However, for readability, these words do not appear in all
uppercase letters in this specification.

All of the text of this specification is normative except sections
explicitly marked as non-normative, examples, and notes. Examples in
this specification are introduced with the words "for example".

Codecs
======

Gzip
----

Codec URI:
    https://purl.org/zarr/spec/codecs/gzip

    
Configuration parameters
~~~~~~~~~~~~~~~~~~~~~~~~

level:
    An integer from 0 to 9 which controls the speed and level of
    compression. A level of 1 is the fastest compression method and
    produces the least compressions, while 9 is slowest and produces
    the most compression. Compression is turned off completely when
    level is 0.

For example, the array metadata below specifies that the compressor is
the Gzip codec configured with a compression level of 1::

    {
        "compressor": {

>             "codec": "https://purl.org/zarr/spec/codecs/gzip",

Reviewer: The spec version isn't part of the URL anymore? We previously had:
```suggestion
            "codec": "https://purl.org/zarr/spec/codecs/gzip/1.0",
```

Reviewer: :+1:

            "configuration": {                                                                                
                "level": 1                                                                                    
            }
        },
    }

    
Format and algorithm
~~~~~~~~~~~~~~~~~~~~

Encoding and decoding is performed using the algorithm defined in
[RFC1951]_.

Encoded data should conform to the Gzip file format [RFC1952]_.

Blosc
-----

Codec URI:
    https://purl.org/zarr/spec/codecs/blosc

    
Configuration parameters
~~~~~~~~~~~~~~~~~~~~~~~~

cname:
    A string identifying the internal compression algorithm to be
    used. At the time of writing, the following values are supported
    by the c-blosc library: "lz4", "lz4hc", "blosclz", "zstd",
    "snappy", "zlib".
    
clevel:

>     An integer between 1 and 9 indicating the compression level.

Reviewer: According to the [c-blosc source file](https://github.com/Blosc/c-blosc/blob/master/blosc/blosc.h#L147-L148), the compression level can also be 0, in which case there is no compression.
I think we should copy/paste the description of the `level` parameter from GZip, as the relation between compression level and speed of compression probably applies to every underlying compression algorithm.
```suggestion
    An integer from 0 to 9 which controls the speed and level of
    compression. A level of 1 is the fastest compression method and
    produces the least compressions, while 9 is slowest and produces
    the most compression. Compression is turned off completely when
    level is 0.
```

shuffle:

>     An integer value in the set {0, 1, 2, -1}. A value of 1
>     indicates that byte-wise shuffling is performed in addition to
>     compression. A value of 2 indicates the bit-wise shuffling is
>     performed in addition to compression. If a value of -1 is given,
>     then default shuffling is used: bit-wise shuffling for buffers
>     with item size of 1 byte, byte-wise shuffling otherwise.
>     Shuffling is turned off completely when the value is 0.

Reviewer: Instead of "in addition to compression", should we say "prior to compression"? It seems more accurate.
We might also mention that shuffling "may improve performance and/or compression ratio" (as stated in https://blosc.org/pages).
```suggestion
    An integer value in the set {0, 1, 2, -1} indicating the way
    bytes or bits are rearranged, which can lead to faster
    and/or greater compression. A value of 1
    indicates that byte-wise shuffling is performed prior to
    compression. A value of 2 indicates the bit-wise shuffling is
    performed prior to compression. If a value of -1 is given,
    then default shuffling is used: bit-wise shuffling for buffers
    with item size of 1 byte, byte-wise shuffling otherwise.
    Shuffling is turned off completely when the value is 0.
```

blocksize:

>     An integer giving the size in bytes of blocks into which a
>     buffer is divided before compression.

Reviewer: Should we mention that when `0` is passed, [an automatic size is used](https://github.com/Blosc/c-blosc/blob/master/blosc/blosc.h#L238-L239)?
```suggestion
    An integer giving the size in bytes of blocks into which a
    buffer is divided before compression. A value of 0
    indicates that an automatic size will be used.
```

For example, the array metadata document below specifies that the
compressor is the Blosc codec configured with a compression level of
1, byte-wise shuffling, the ``lz4`` compression algorithm and the
default block size::

    {
        "compressor": {

>             "codec": "https://purl.org/zarr/spec/codecs/blosc",

Reviewer: The spec version isn't part of the URL anymore? We previously had:
```suggestion
            "codec": "https://purl.org/zarr/spec/codecs/blosc/1.0",
```

Author: Yes, see above.

Note that Blosc 2 would be treated as an entirely new codec, and we would probably give it a URI like `https://purl.org/zarr/spec/codecs/blosc2`.

Reviewer: :+1:

                "clevel": 1,
                "shuffle": 1,
                "blocksize": 0
            }
        },
    }

Format and algorithm
~~~~~~~~~~~~~~~~~~~~

Blosc is a meta-compressor, which divides an input buffer into blocks,
then applies an internal compression algorithm to each block, then
packs the encoded blocks together into a single output buffer with a
header. The format of the encoded buffer is defined in [BLOSC]_. The
reference implementation is provided by the `c-blosc library
<https://github.com/Blosc/c-blosc>`_.

Deprecated codecs
=================

There are no deprecated codecs at this time.

References
==========

.. [RFC2119] S. Bradner. Key words for use in RFCs to Indicate
   Requirement Levels. March 1997. Best Current Practice. URL:
   https://tools.ietf.org/html/rfc2119

.. [RFC1951] P. Deutsch. DEFLATE Compressed Data Format Specification version
   1.3. Requirement Levels. May 1996. Informational. URL:
   https://tools.ietf.org/html/rfc1951

.. [RFC1952] P. Deutsch. GZIP file format specification version 4.3.
   Requirement Levels. May 1996. Informational. URL:
   https://tools.ietf.org/html/rfc1952

.. [BLOSC] F. Alted. Blosc Chunk Format. URL:

>    https://github.com/Blosc/c-blosc/blob/master/README_CHUNK_FORMAT.rst

Reviewer: This document describes the "Blosc Chunk Format". Does it mean a Zarr chunk consists of one or more Blosc chunks? If so, is the Zarr chunk a simple concatenation of the Blosc chunks?

Author: The "Blosc Chunk Format" describes how blosc encodes an input buffer. So one zarr chunk becomes one blosc chunk once encoded.

Reviewer: :+1:
